### PR TITLE
Ansible add metrics

### PR DESCRIPTION
**Pull request by asmacdo**
**Description of the change:**

**Motivation for the change:**

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [ ] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [ ] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

### CODE REVIEW

File: internal/ansible/apiserver/apiserver.go
Comment (jmrodri) [inline internal/ansible/apiserver/apiserver.go:41]:
```suggestion
		log.Info("apiserver has received a POST")
```

File: internal/ansible/metrics/metrics.go
Comment (jmrodri) [inline internal/ansible/metrics/metrics.go:122]: Are we keeping "at least this worked 1" in the code?

File: internal/ansible/metrics/metrics.go
Comment (jmrodri) [inline internal/ansible/metrics/metrics.go:126]:
```suggestion

```

File: internal/ansible/metrics/metrics.go
Comment (jmrodri) [inline internal/ansible/metrics/metrics.go:128]:
```suggestion

```

File: internal/ansible/metrics/metrics.go
Comment (jmrodri) [inline internal/ansible/metrics/metrics.go:135]:
```suggestion

```

File: internal/ansible/metrics/metrics.go
Comment (jmrodri) [inline internal/ansible/metrics/metrics.go:155]:
```suggestion

```

File: internal/ansible/metrics/metrics.go
Comment (jmrodri) [inline internal/ansible/metrics/metrics.go:164]: Should these be `Info` level?

File: internal/ansible/metrics/metrics.go
Comment (jmrodri) [inline internal/ansible/metrics/metrics.go:168]: Should this be an error?

**Review by jmrodri — CHANGES_REQUESTED** [review]
I think the logs need to get cleaned up before merging. And I found a typo as well.

File: internal/ansible/metrics/metrics.go
Comment (asmacdo) [reply internal/ansible/metrics/metrics.go]: nope :)

File: internal/ansible/apiserver/apiserver.go
Comment (fabianvf) [inline internal/ansible/apiserver/apiserver.go:30]: This struct should probably have options like the port/listen address

File: internal/ansible/apiserver/apiserver.go
Comment (fabianvf) [inline internal/ansible/apiserver/apiserver.go:55]: These can probably be debug messages

File: internal/ansible/apiserver/apiserver.go
Comment (fabianvf) [inline internal/ansible/apiserver/apiserver.go:63]: same here

File: internal/ansible/metrics/metrics.go
Comment (fabianvf) [inline internal/ansible/metrics/metrics.go:156]: is metricSpec.Name supposed to be the actual value of metricSpec.Name?

File: internal/cmd/ansible-operator/run/cmd.go
Comment (fabianvf) [inline internal/cmd/ansible-operator/run/cmd.go:267]: With the options included in the metric we could the listen address and port here instead, which I think would be more clear for future maintainers

**Review by fabianvf — CHANGES_REQUESTED** [review]

File: internal/ansible/metrics/metrics.go
Comment (asmacdo) [reply internal/ansible/metrics/metrics.go]: This can't happen, just a debug sanity check

File: internal/ansible/metrics/metrics.go
Comment (ryantking) [inline internal/ansible/metrics/metrics.go:169]: Can we move this (and the other branches) to helpers?
```go
case prometheus.Gauge:
    return incermentUserGauge(/* ... */)
// etc
```

File: internal/ansible/metrics/metrics.go
Comment (ryantking) [inline internal/ansible/metrics/metrics.go:192]: Here's my idea, for a refactor of this function that works around the `prometheus.Summary` and `prometheus.Histogram` interface collision. It makes some of my prometheus comments null if you go with it.
```go
func HandleUserMetric(r prometheus.Registerer, metricSpec UserMetric) error {
    if err := validateMetricSpec(metricSpec); err != nil { /* ... */ } // one and only one collector is configured
    collector, err := ensureCollector(metricSpec) // get or create then return collector
    if err != nil { /* ... */ }
    if metricSpec.Counter != nil {
        return handleUserCounter(metricSpec)
    } else if metricSpec.Guage != nil {
        // etc
    }

    return errors.New("...")
}

func handleUserCounter(collector prometheus.Collector, metricSpec UserMetric) error {
    v := collector.(prometheus.Counter) // should probably check the assertion
    if metricSpec.Counter == nil {
 		return fmt.Errorf("cannot change metric type of %s, which is a counter", metricSpec.Name)
 	}
 	if metricSpec.Counter.Inc {
 		v.Inc()
 	} else if metricSpec.Counter.Add != 0.0 {
 		v.Add(metricSpec.Counter.Add)
 	}
    return nil
```

Comment (ryantking) [reply internal/ansible/metrics/metrics.go:192]: Actually, you definitely need to check the assertion. If you POST the same metric name with two different types it'll blow up

File: internal/ansible/metrics/metrics.go
Comment (asmacdo) [reply internal/ansible/metrics/metrics.go:207]: +1 This shouldn't ever occur when using with the operator_sdk.util ansible collection, but best to be robust.

File: internal/ansible/apiserver/apiserver.go
Comment (ryantking) [inline internal/ansible/apiserver/apiserver.go:70]: This function as it stands could blast through file descriptors. You need to add to the top:
```go
defer func() {
    io.Copy(io.Discard, r.Body)
    r.Body.Close()
}()
```

This will ensure that the request body gets closed and all the data gets read from it. In go, it's not enough to just close the request body (or response body when doing requests), you must also call something like `io.Copy` or `ioutil.ReadAll` in order to read all the bytes from it. If there is still data on the connection or it's not closed, go will maintain an open connection (file descriptor) while the process is alive.

File: internal/ansible/metrics/metrics.go
Comment (ryantking) [inline internal/ansible/metrics/metrics.go:240]: Instead of the type switch, you can do:
```
switch {
case metricSpec.Gauge != nil:
    // gauge stuff
case metricSpec.Counter != nil:
    // counter stuff
// etc
}
```
Not a blocker for this PR, but would allow you to split the summary and histogram helper functions.

File: internal/cmd/ansible-operator/run/cmd.go
Comment (ryantking) [inline internal/cmd/ansible-operator/run/cmd.go:272]: Mislabled, `server` is actually an error in this case.

**Review by ryantking — CHANGES_REQUESTED** [review]
Two small fixes then LGTM

**Review by fabianvf — APPROVED** [review]
lgtm

File: internal/ansible/metrics/metrics.go
Comment (asmacdo) [reply internal/ansible/metrics/metrics.go:240]: I dont think that will work in edge cases where the user is erroneously mixing types. For example, if the user creates a metric "my_counter_metric", and then calls the gauge function on "my_counter_metric". In this case, the current code will see that it is a counter, and handle it as a counter, and send a failure message that "you cant change counter to gauge". For your suggestion, it will attempt to handle it as a gauge, which could cause silent problems I think?

Comment (ryantking) [reply internal/ansible/metrics/metrics.go:240]: Gotcha, do we validate anywhere else that for existing metrics, the type doesn't change?

File: internal/ansible/apiserver/apiserver.go
Comment (asmacdo) [reply internal/ansible/apiserver/apiserver.go]: Makes sense. Nice catch, TIL :)

File: internal/ansible/metrics/metrics.go
Comment (asmacdo) [reply internal/ansible/metrics/metrics.go:243]: We didn't collect custom metrics from users, so AFAIK it hasn't come up before.

**Review by ryantking — APPROVED** [review]
lgtm

**Pull request merged**
